File: go/library/library.go
package library

import (
	pb "grpc-vimn-demo-go/protos"
	"strings"
)

var authors = []*pb.Author{
	{Id: 1, Name: "John Ronald Reuel Tolkien"},
	{Id: 2, Name: "William Golding"},
	{Id: 3, Name: "William Shakespeare"},
}

var books = []*pb.Book{
	{Id: 1, Title: "The Hobbit", Authors: []*pb.Author{authors[0]}},
	{Id: 2, Title: "The Lord Of The Rings", Authors: []*pb.Author{authors[0]}},
	{Id: 3, Title: "Lord Of The Flies", Authors: []*pb.Author{authors[1]}},
	{Id: 4, Title: "Hamlet", Authors: []*pb.Author{authors[2]}},
}

func ListBooks() []*pb.Book {
	return books
}

func FilterBooks(title string, authorName string) []*pb.Book {
	books := books

	if title != "" {
		filtered := make([]*pb.Book, 0)
		for _, book := range books {
			if strings.Contains(strings.ToLower(book.Title), strings.ToLower(title)) {
				filtered = append(filtered, book)
			}
		}
		books = filtered
	}

	if authorName != "" {
		filtered := make([]*pb.Book, 0)
	bookLoop:
		for _, book := range books {
			for _, author := range book.Authors {
				if strings.Contains(strings.ToLower(author.Name), strings.ToLower(authorName)) {
					filtered = append(filtered, book)
					continue bookLoop
				}
			}
		}
		books = filtered
	}

	return books
}

File: python/library.py
from typing import List, Optional

import library_pb2


AUTHORS = [
    library_pb2.Author(id=1, name="John Ronald Reuel Tolkien"),
    library_pb2.Author(id=2, name="William Golding"),
    library_pb2.Author(id=3, name="William Shakespeare"),
]

BOOKS = [
    library_pb2.Book(id=1, title="The Hobbit", authors=[AUTHORS[0]]),
    library_pb2.Book(id=2, title="The Lord Of The Rings", authors=[AUTHORS[0]]),
    library_pb2.Book(id=3, title="Lord Of The Flies", authors=[AUTHORS[1]]),
    library_pb2.Book(id=4, title="Hamlet", authors=[AUTHORS[2]]),
]


def list_books() -> List[library_pb2.Book]:
    return [b for b in BOOKS]


def filter_books(
    title: Optional[str] = None,
    author: Optional[str] = None,
) -> List[library_pb2.Book]:
    books = [b for b in BOOKS]

    if title:
        books = [b for b in books if title.lower() in b.title.lower()]

    if author:
        books = [b for b in books if any(author.lower() in a.name.lower() for a in b.authors)]

    return books
